specs/exp-proof.md

```diff
@@ -71,6 +72,11 @@ The exponentiation circuit consists of the following columns:
     - Both multiplicand's of the `mul_gadget`, i.e. `a` and `b` MUST equal `base` of the exponentiation operation, that is:
         - `mul_gadget.a == base`
         - `mul_gadget.b == base` (for both these cases we equate each 64-bit limb)
+- For every row where `padding == true`, validate that:
+    - next row is also `padding == true`
```

Comment:
This will cause that we can never stop padding. We can't force that the next row is `padding` nor anything as the algorithm has to end at some point.

Reply:
Maybe we could include another column that indicates if its the last row in padding?

Reply:
As you see [here](https://github.com/scroll-tech/zkevm-specs/blob/exp-padding/src/zkevm_specs/exp_circuit.py#L96-L101), each time we verify a step, we pass 2 consecutive rows from the exponentiation table. Which means, when `row[i]` is the last row, `row[(i+1) % n]` is in fact again the first row.

We have a `is_last` column to mark whether or not the row is the last step within a particular exponentiation trace. [Read more here](https://github.com/scroll-tech/zkevm-specs/blob/exp-padding/specs/exp-proof-design-doc.md) and [here](https://github.com/scroll-tech/zkevm-specs/blob/exp-padding/specs/exp-proof.md).

In addition to that, you can also add a `is_final` advice column that stands for the last row in the entire circuit. So then you could use this in your padding logic, i.e. something like:
```
- for every row where `padding == True` and `is_final == False`:
    - next row is also `padding == True`
    - all other columns are set to 0
```

You could have additional constraint to check that `is_final` advice column is set to `1` only once throughout the circuit.

@CPerezz What do you think?

Reply:
@roynalnaruto can this [constraint](https://github.com/scroll-tech/zkevm-specs/blob/master/src/zkevm_specs/exp_circuit.py#L34) be deleted? Because we are enabling the padding rows now

Reply:
Agree completelly with @roynalnaruto . Ideally we won't add an extra column. But it;s fine if we don't see any other way.

Reply:
@rrzhang139 all rows will be usable, so yes, you can modify/add/remove constraints as you may feel right. In this particular case (that [you have linked](https://github.com/scroll-tech/zkevm-specs/blob/master/src/zkevm_specs/exp_circuit.py#L32-L33)), we wouldn't need this anymore.